### PR TITLE
fix the race condition in test logger

### PR DESCRIPTION
**Pull request by BinacsLee**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

When using test logger in an unsynchronized goroutine, we will meet a race condition such as: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/93/pull-kueue-test-unit-main/1499496533626720256 (in https://github.com/kubernetes-sigs/kueue/pull/93)

We have to synchronize the goroutines in the unit test to solve the problem.

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

/assign @alculquicondor 
/cc @ahg-g

### CODE REVIEW

**Comment by alculquicondor** [comment]
Can you summarize how you concluded that this is the problem? And what experiments did you conduct to verify that this fixes the problem.

The logger itself is thread-safe. So I was thinking that the problem might be when trying to access a value passed to the logger.

**Comment by BinacsLee** [comment]
> Can you summarize how you concluded that this is the problem? And what experiments did you conduct to verify that this fixes the problem?
> 
> The logger itself is thread-safe. So I was thinking that the problem might be when trying to access a value passed to the logger.

hi @alculquicondor !

First, I tried to reproduce the issue with `go test ./pkg/scheduler/... -race -count 1000`. Then I found that sometimes I can see the "Log in goroutine after..." message when a race condition is generated.

We can found that this panic is generated in https://github.com/golang/go/blob/63bd6f68e6cbb237b46a99775103758afaee370a/src/testing/testing.go#L887 and with the help of [related comments](https://github.com/golang/go/blob/63bd6f68e6cbb237b46a99775103758afaee370a/src/testing/testing.go#L1424) we can conclude that **it will be illegal to use `test.logger` in a test case that has already completed**.

That's to say, before executing the next test case, we have to make sure that no test.logger is still being used anywhere.

By adding a channel to synchronize the goroutines, we can ensure that test.logger will only be used in each individual testcase *(no two testcases are using test.logger at the same time, especially if one of them is considered complete **even though it still has a goroutine running**)*.

We found the cause of the problem, and we can solve it directly through coroutine synchronization. I verified the above through multiple test commands as well.

**Comment by alculquicondor** [comment]
Thanks for the explanation. That makes sense.

However, I think we should implement it differently. Use `sync.WaitGroup` and do not use `wantScheduled` as the number of routines we are waiting for. If we introduce a bug in the scheduler, the number of goroutines would be different from `wantScheduled`. Instead, do `wg.Add` everytime that a routine is launched.

**Comment by BinacsLee** [comment]
Sound great to use `sync.WaitGroup` instead of `channel`. And I have updated the code.

In addition, I personally think it is worthwhile to create a function level logger, which can play a role in many similar scenarios. This function logger is located in pkg/util/function/logger.go . What is your opinion please?

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -36,22 +36,25 @@ import (
...
 type Scheduler struct {
-	queues   *queue.Manager
-	cache    *cache.Cache
-	client   client.Client
-	recorder record.EventRecorder
+	queues     *queue.Manager
+	cache      *cache.Cache
+	client     client.Client
+	recorder   record.EventRecorder
+	funcLogger function.Logger
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:48]: `admissionRoutineWrapper`

but I don't see anything wrong just having a pointer to a `WaitGroup` here instead. Then, it would be called `admissionRoutineWG`.

File: pkg/util/function/logger.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package function
+
+// Logger is used to log function.
```
Comment (alculquicondor) [inline pkg/util/function/logger.go:3]: I wonder if you have seen this pattern somewhere else.

`Logger` doesn't seem like the correct name (it's not logging itself). Maybe I would call it `Wrapper` (`function.Wrapper` when you include the package name).

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (BinacsLee) [reply pkg/scheduler/scheduler.go]: Using a pointer to `WaitGroup` is absolutely right... and I just want to make it more extensible by using a `wrapper` cause only in unit tests do we need this `WaitGroup`.

File: pkg/util/function/logger.go (hunk 2 above)
Comment (BinacsLee) [reply pkg/util/function/logger.go]: Thank you very much for pointing out my mistake. More precisely, it's more like a wrapper, so wrapper is the best name.

If you think this wrapper is useful, I will try to use it to fix other possible goroutine leaks.

**Comment by alculquicondor** [comment]
Sounds good, it might be useful long term.

I like the `routine.Wrapper` name.

/lgtm
/approve
